### PR TITLE
ModuleInstaller.Upgrade now properly handles installing deps.

### PR DESCRIPTION
**Pull request by pjf**
The upgrade command now properly handles installing dependencies
to satisfy constraints. It was always supposed to do this (which is
why we use the relationship resolver to figure out what needs to
be installed), but the previous code didn't like trying to "upgrade"
a module that wasn't previously installed.

Removes urgency for KSP-CKAN/CKAN#809.
Brings great joy to KSP-CKAN/NetKAN#889.

Sample output using the RSS split:

```
$ ckan.exe upgrade RealSolarSystem

Upgrading modules...

3484 [1] WARN CKAN.ModuleInstaller (null) - RealSolarSystem is already
at the latest version, reinstalling
Installing previously uninstalled mod RSS-Core

Done!
```

Attenioning @Dazpoet, because hopefully this is smile-bringing. :)

### CODE REVIEW

**Comment by pjf** [comment]
Note that this *does not* change the GUI behaviour. It doesn't use the `ModuleIntaller.Upgrade()` interface. That's really something I'd love to fix, but it's going to be much more tricky than this humble PR.

**Comment by Dazpoet** [comment]
Oooh does this mean that we can tell users to solve problems with updated dependencies with a simple CLI command, because it looks like that's exactly what you're saying and right now nothing could make me happier!

**Comment by pjf** [comment]
> Oooh does this mean that we can tell users to solve problems with updated dependencies with a simple CLI command, because it looks like that's exactly what you're saying and right now nothing could make me happier!

If we're talking about users who have mods installed, and those mods now have new dependencies that need to be installed as part of an upgrade, then the answer is yes. :)

**Comment by Dazpoet** [comment]
That is exactly what I meant :+1:

What about a case where a dependency was removed instead, I must admit I don't know if those currently throws a Kraken but handling the situation in a good way would still be advisable. A case where a mod recently removed a dependency would be SETIBalandeMod which removed ProceduralParts when going 0.8.7->0.8.8

**Comment by pjf** [comment]
> What about a case where a dependency was removed instead, I must admit I don't know if those currently throws a Kraken but handling the situation in a good way would still be advisable. A case where a mod recently removed a dependency would be SETIBalandeMod which removed ProceduralParts when going 0.8.7->0.8.8

The way in which this works (and always always meant to work) is that it takes the list of mods we're asked to upgrade, hands it to the relationship resolver indicating what dependencies are required (but not already installed), and then upgrades/installs the resulting list. We'll never try to un-install a mod because of this, so in the case of a removed dependency the mod that was depends upon sticks around. If the mod sticks around results in a conflict, then we should fail our sanity checks and bail out.

In the case of a mod going from a dependency to a conflict for the same mod, we'll probably need to implement KSP-CKAN/CKAN#809 (extended upgrade syntax), or a relationship resolver that can suggest mod removals, along with some way of confirming with the user that we're going to un-install a mod as part of the upgrade.

**TL;DR:** If a mod removes a dependency, we'll just leave it installed.

**Comment by Dazpoet** [comment]
Thanks for clearing that out :+1: 

Unfortunaltly I feel my code-understanding-fo is to low to claim to have reviewed this prior to a merge so I hope someone who understands it can fix that part.

**Comment by pjf** [comment]
I'm in my code, merging my own pulls. ;)
